File: backend/app/schemas/institution_student.py
"""Pydantic schemas for the Institution Student Directory
(backend/app/api/institution.py, GET /institution/students,
GET /institution/students/{student_id}).

Every field here is backed by a real column or a real, already-derived
value -- nothing is fabricated. In particular:
  - `placement_status` reuses the EXACT same three-bucket definition as
    the Institution Dashboard (institution_service._placement_buckets):
    PLACED (>=1 SELECTED application), APPLYING (>=1 application, none
    SELECTED), NOT_PARTICIPATING (no applications at all). The dashboard
    and this directory must never define "placed" differently -- see
    institution_student_service.py's own docstring.
  - `internship_status` is the same three-bucket logic narrowed to
    opportunity_type = 'INTERNSHIP' only.
  - There is no `eligible` field anywhere here -- same reasoning as
    InstitutionOverviewResponse.eligibility_note (no eligibility data
    model exists in this schema).
  - No email, phone, or any other sensitive profile field is exposed --
    see institution_student_service.py and migration 039's own comments.
"""

from typing import Literal

from pydantic import BaseModel

PlacementStatus = Literal["PLACED", "APPLYING", "NOT_PARTICIPATING"]
InternshipStatus = Literal["SELECTED", "APPLYING", "NONE"]


class StudentSkillSummary(BaseModel):
    skill_name: str
    proficiency_level: str
    is_verified: bool


class StudentSummary(BaseModel):
    """One row in the Student Directory list."""

    id: str
    full_name: str | None
    username: str | None
    avatar_url: str | None
    # institution_link_requests.id for this student's current APPROVED
    # link -- lets the frontend call the existing unlink endpoint
    # (POST /institution-links/{id}/unlink, migration 038) directly from
    # the directory/detail views without a second lookup. None only in
    # the (should-never-happen) case where a student_profiles row has
    # institution_id set but no matching APPROVED request is found.
    link_request_id: str | None
    # None means "Unassigned" -- see database/migrations/
    # 040_institution_departments.sql. `department` is the resolved
    # display name ("Unassigned" when department_id is None), NOT the
    # student's own free-text student_profiles.department value anymore.
    department_id: str | None
    department: str
    # student_profiles.graduation_year -- displayed as "Batch".
    batch: int | None
    cgpa: float | None
    percentage: float | None
    placement_status: PlacementStatus
    internship_status: InternshipStatus
    # Up to 5 skills, highest-proficiency first -- not the full list (see
    # get_student_detail for that). Matches the dashboard's own
    # top-skills cap philosophy (institution_service._TOP_SKILLS_LIMIT).
    top_skills: list[str]
    # Same algorithm as frontend/lib/student/profile.ts's
    # getProfileCompletion -- ported here so the institution-visible
    # figure and the student's own never silently drift apart. See
    # institution_student_service._profile_completion.
    profile_completion: int


class StudentListSummary(BaseModel):
    """Roster-wide counts (before search/filter/pagination is applied) --
    reuses the exact same `_placement_buckets` computation as the
    Institution Dashboard (institution_service.compute_institution_overview),
    so these summary cards and the Dashboard's own KPIs can never disagree
    for the same institution."""

    total_students: int
    placed: int
    unplaced: int
    no_applications: int
    internship_selected: int


class DepartmentOption(BaseModel):
    """One of the institution's real departments (id + display name) --
    NOT a free-text value. Includes inactive departments too, so a
    student historically assigned to a since-deactivated department can
    still be found via this filter (Part 12/23 of this phase: inactive
    departments must not disappear from historical student data)."""

    id: str
    name: str


class StudentListFilters(BaseModel):
    """Distinct values available across the institution's FULL linked
    roster (not just the current page) -- lets the frontend populate
    filter dropdowns without a second round trip or guessing options
    from one page's worth of rows. "Unassigned" is not listed here (it's
    a fixed, well-known filter value the frontend already knows about,
    not a real department row) -- pass `department=unassigned` to filter
    for it."""

    departments: list[DepartmentOption]
    batches: list[int]


class StudentListResponse(BaseModel):
    students: list[StudentSummary]
    total: int
    page: int
    page_size: int
    filters: StudentListFilters
    summary: StudentListSummary


class StudentProjectSummary(BaseModel):
    id: str
    title: str
    description: str | None
    project_url: str | None
    repo_url: str | None
    is_ongoing: bool
    skills: list[str]


class StudentCertificationSummary(BaseModel):
    id: str
    name: str
    issuing_organization: str | None
    issue_date: str | None
    credential_url: str | None


class StudentAchievementSummary(BaseModel):
    id: str
    title: str
    description: str | None
    achievement_date: str | None
    issuing_organization: str | None


class StudentApplicationSummary(BaseModel):
    id: str
    opportunity_type: str  # "INTERNSHIP" | "JOB"
    opportunity_title: str | None
    company_name: str | None
    status: str
    applied_at: str | None


class StudentAssessmentSummary(BaseModel):
    assessment_title: str | None
    skill_name: str | None
    status: str
    percentage: float | None
    submitted_at: str | None


class StudentInterviewSummary(BaseModel):
    """Deliberately excludes `notes` (industry-private preparation notes,
    030_industry_interviews.sql's own comment: "never exposed to any
    other role") -- see institution_student_service.py."""

    id: str
    opportunity_title: str | None
    scheduled_at: str
    mode: str
    status: str


class StudentDetailResponse(BaseModel):
    id: str
    full_name: str | None
    username: str | None
    avatar_url: str | None
    # See StudentSummary.link_request_id's own docstring.
    link_request_id: str | None

    # ---- academics ----
    department_id: str | None
    department: str
    # The free-text value from the student's OWN /student/profile page --
    # informational only, shown so the institution can decide which real
    # department to assign; never the authoritative value itself.
    self_reported_department: str | None
    batch: int | None
    degree: str | None
    cgpa: float | None
    percentage: float | None
    profile_completion: int

    # ---- skills ----
    skills: list[StudentSkillSummary]

    # ---- portfolio (read-only; no new schema, existing student_* tables) ----
    projects: list[StudentProjectSummary]
    certifications: list[StudentCertificationSummary]
    achievements: list[StudentAchievementSummary]

    # ---- applications / placement / internships ----
    applications: list[StudentApplicationSummary]
    placement_status: PlacementStatus
    internship_status: InternshipStatus

    # ---- assessments ----
    assessments_completed: int
    average_assessment_percentage: float | None
    assessments: list[StudentAssessmentSummary]

    # ---- interviews ----
    interviews: list[StudentInterviewSummary]

    # Always present -- explains what is NOT shown and why (no resume
    # storage, no GitHub/LinkedIn field, no eligibility/backlog data).
    notes: list[str]

File: frontend/types/institution-student.ts
// Mirrors backend/app/schemas/institution_student.py -- the Institution
// Student Directory (GET /institution/students,
// GET /institution/students/{id}).
//
// `placement_status` reuses the EXACT same three-bucket definition as
// the Institution Dashboard: PLACED (>=1 SELECTED application), APPLYING
// (>=1 application, none SELECTED), NOT_PARTICIPATING (no applications).
// The Dashboard and this Directory must never disagree about what
// "placed" means -- see the backend service's own docstring.

export type PlacementStatus = "PLACED" | "APPLYING" | "NOT_PARTICIPATING";
export type InternshipStatus = "SELECTED" | "APPLYING" | "NONE";

export const PLACEMENT_STATUS_LABELS: Record<PlacementStatus, string> = {
  PLACED: "Placed",
  APPLYING: "Unplaced",
  NOT_PARTICIPATING: "No Applications",
};

export const INTERNSHIP_STATUS_LABELS: Record<InternshipStatus, string> = {
  SELECTED: "Selected",
  APPLYING: "Applying",
  NONE: "None",
};

export interface StudentSummary {
  id: string;
  full_name: string | null;
  username: string | null;
  avatar_url: string | null;
  link_request_id: string | null;
  /** null means "Unassigned" -- see types/institution-department.ts.
   * `department` is the resolved display name ("Unassigned" when
   * department_id is null), not free text anymore. */
  department_id: string | null;
  department: string;
  batch: number | null;
  cgpa: number | null;
  percentage: number | null;
  placement_status: PlacementStatus;
  internship_status: InternshipStatus;
  top_skills: string[];
  profile_completion: number;
}

/** One of the institution's real departments (id + name) -- includes
 * inactive ones too, so a student historically assigned to a
 * since-deactivated department can still be filtered for. "Unassigned"
 * is not listed here -- it's a fixed, well-known filter value
 * (UNASSIGNED_DEPARTMENT_FILTER below), not a real department row. */
export interface DepartmentOption {
  id: string;
  name: string;
}

export const UNASSIGNED_DEPARTMENT_FILTER = "unassigned";

export interface StudentListFilters {
  departments: DepartmentOption[];
  batches: number[];
}

/** Roster-wide counts (before search/filter/pagination) -- the same
 * `_placement_buckets` computation as the Institution Dashboard, so
 * these summary cards can never disagree with the Dashboard's own KPIs
 * for the same institution. */
export interface StudentListSummary {
  total_students: number;
  placed: number;
  unplaced: number;
  no_applications: number;
  internship_selected: number;
}

export interface StudentListResponse {
  students: StudentSummary[];
  total: number;
  page: number;
  page_size: number;
  filters: StudentListFilters;
  summary: StudentListSummary;
}

export interface StudentSkillSummary {
  skill_name: string;
  proficiency_level: string;
  is_verified: boolean;
}

export interface StudentProjectSummary {
  id: string;
  title: string;
  description: string | null;
  project_url: string | null;
  repo_url: string | null;
  is_ongoing: boolean;
  skills: string[];
}

export interface StudentCertificationSummary {
  id: string;
  name: string;
  issuing_organization: string | null;
  issue_date: string | null;
  credential_url: string | null;
}

export interface StudentAchievementSummary {
  id: string;
  title: string;
  description: string | null;
  achievement_date: string | null;
  issuing_organization: string | null;
}

export interface StudentApplicationSummary {
  id: string;
  opportunity_type: "INTERNSHIP" | "JOB";
  opportunity_title: string | null;
  company_name: string | null;
  status: string;
  applied_at: string | null;
}

export interface StudentAssessmentSummary {
  assessment_title: string | null;
  skill_name: string | null;
  status: string;
  percentage: number | null;
  submitted_at: string | null;
}

export interface StudentInterviewSummary {
  id: string;
  opportunity_title: string | null;
  scheduled_at: string;
  mode: string;
  status: string;
}

export interface StudentDetail {
  id: string;
  full_name: string | null;
  username: string | null;
  avatar_url: string | null;
  link_request_id: string | null;

  department_id: string | null;
  department: string;
  /** The student's own free-text student_profiles.department value --
   * informational only, shown to help decide which real department to
   * assign; never the authoritative value. */
  self_reported_department: string | null;
  batch: number | null;
  degree: string | null;
  cgpa: number | null;
  percentage: number | null;
  profile_completion: number;

  skills: StudentSkillSummary[];

  projects: StudentProjectSummary[];
  certifications: StudentCertificationSummary[];
  achievements: StudentAchievementSummary[];

  applications: StudentApplicationSummary[];
  placement_status: PlacementStatus;
  internship_status: InternshipStatus;

  assessments_completed: number;
  average_assessment_percentage: number | null;
  assessments: StudentAssessmentSummary[];

  interviews: StudentInterviewSummary[];

  notes: string[];
}

export interface StudentListParams {
  search?: string;
  department?: string;
  batch?: number;
  cgpa_min?: number;
  cgpa_max?: number;
  placement_status?: PlacementStatus;
  internship_status?: InternshipStatus;
  skill?: string;
  sort_by?: "name" | "cgpa" | "department" | "placement_status";
  sort_dir?: "asc" | "desc";
  page?: number;
  page_size?: number;
}
